CMakeLists.txt

```diff
@@ -0,0 +1,79 @@
+set(SG14_TEST_SOURCE_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}/SG14_test")
+
+# Output binary to predictable location.
+set(BINARY_OUT_DIR ${CMAKE_CURRENT_BINARY_DIR}/bin)
+set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${BINARY_OUT_DIR})
+set(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${BINARY_OUT_DIR})
+set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY ${BINARY_OUT_DIR})
+
+foreach(OUTPUTCONFIG ${CMAKE_CONFIGURATION_TYPES})
+	string(TOUPPER ${OUTPUTCONFIG} OUTPUTCONFIG)
+	set(CMAKE_RUNTIME_OUTPUT_DIRECTORY_${OUTPUTCONFIG} ${BINARY_OUT_DIR})
+	set(CMAKE_LIBRARY_OUTPUT_DIRECTORY_${OUTPUTCONFIG} ${BINARY_OUT_DIR})
+	set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY_${OUTPUTCONFIG} ${BINARY_OUT_DIR})
+endforeach(OUTPUTCONFIG CMAKE_CONFIGURATION_TYPES)
+
+link_directories(${CMAKE_CURRENT_BINARY_DIR}/lib) # For future use with testing library.
+
+##
+# Project
+##
+file(GLOB_RECURSE INCLUDES "${SG14_INCLUDE_DIRECTORY}/*.hpp" "${SG14_INCLUDE_DIRECTORY}/*.h" "${SG14_INCLUDE_DIRECTORY}/*.tpp")
+add_library(${PROJECT_NAME} INTERFACE)
+target_include_directories(${PROJECT_NAME} INTERFACE ${SG14_INCLUDE_DIRECTORY})
+target_sources(${PROJECT_NAME} INTERFACE
+	$<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>
+	$<BUILD_INTERFACE:${INCLUDES}>
+)
+target_include_directories(${PROJECT_NAME} INTERFACE
+	$<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>
+	$<BUILD_INTERFACE:${SG14_INCLUDE_DIRECTORY}>
+)
+
+##
+# Unit Tests
+##
+set(TEST_SOURCE_FILES
+    ${SG14_TEST_SOURCE_DIRECTORY}/main.cpp
+    ${SG14_TEST_SOURCE_DIRECTORY}/inplace_function_test.cpp
+    ${SG14_TEST_SOURCE_DIRECTORY}/plf_colony_test.cpp
+    ${SG14_TEST_SOURCE_DIRECTORY}/ring_test.cpp
+    ${SG14_TEST_SOURCE_DIRECTORY}/slot_map_test.cpp
+    ${SG14_TEST_SOURCE_DIRECTORY}/transcode_test.cpp
+    ${SG14_TEST_SOURCE_DIRECTORY}/uninitialized_test.cpp
+    ${SG14_TEST_SOURCE_DIRECTORY}/unstable_remove_test.cpp
+)
+
+set(TEST_NAME ${PROJECT_NAME}_tests)
+add_executable(${TEST_NAME} ${TEST_SOURCE_FILES})
+target_link_libraries(${TEST_NAME} ${PROJECT_NAME} ${CMAKE_THREAD_LIBS_INIT})
+target_include_directories(${TEST_NAME} PRIVATE "${SG14_TEST_SOURCE_DIRECTORY}")
+
+# Compile options
+if ("${CMAKE_CXX_COMPILER_ID}" MATCHES "Clang")
+	target_compile_options(${TEST_NAME} PRIVATE -Wall -Wextra -Wfatal-errors -Werror)
+	set_source_files_properties(${SG14_TEST_SOURCE_DIRECTORY}/plf_colony_test.cpp PROPERTIES
+		COMPILE_FLAGS "-Wno-unused-parameter"
+	)
+elseif ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "GNU")
+	target_compile_options(${TEST_NAME} PRIVATE -Wall -Wextra -Wfatal-errors -Werror)
+	set_source_files_properties(${SG14_TEST_SOURCE_DIRECTORY}/plf_colony_test.cpp PROPERTIES
+		COMPILE_FLAGS "-Wno-unused-parameter"
+	)
+elseif ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Intel")
+  # using Intel C++
+elseif ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "MSVC")
+	set_property(DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR} PROPERTY VS_STARTUP_PROJECT ${TEST_NAME})
+	target_compile_options(${TEST_NAME} PRIVATE /Zc:__cplusplus /permissive- /W4)# /WX) temporarily allow warnings
```

Comment:
> This does cause a rather banal compile issue in colony, and a few warnings (that should be looked at) in slot_map.

Can you copy and paste the error messages into your commit message? That'll help the next person clean them up, and/or just satisfy the reader's curiosity as to what _are_ these mysterious error messages.  Cf. https://github.com/zenorogue/hyperrogue/commit/739146246c5ae1f0a13600bef4b882f9b291465d

Reply:
Or should I just add the fix? With details in the commit message.

Reply:
I believe 8d356b7 will fix the MSVC warnings in `slot_map.h`.

The two in `transcode.h` would be easy to fix with a cast, but `transcode.h` is really just a stub, hasn't been touched since its initial commit in October 2016, and I wonder if it's time to just delete it from the master branch. @hatcat may have an opinion.

Reply:
Looks like it. I had planned to fix the warnings after this PR. If there are any left I'll take a look at minimum.